Enhancement: #218 Fuzzy search for QuickOpen

activcoding opened the pull request:
### Description

This PR adds a new file, "FuzzySearch.swift", to the project. The file provides fuzzy search functionality for searching an array of URLs based on a query string. Also modified "QuickOpenViewModel.swift" to integrate the fuzzy search feature using the new 'search' function.

### Related Issues

* Closes #218 

### Checklist

- [x] I read and understood the [contributing guide](https://github.com/CodeEditApp/CodeEdit/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/CodeEditApp/CodeEdit/blob/main/CODE_OF_CONDUCT.md)
- [x] The issues this PR addresses are related to each other
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] My changes are all related to the related issue above
- [x] I documented my code
- [x] Implement the Levenshtein-Algorithm
- [x] performance testing to validate the efficiency of the fuzzy search algorithm
- [x] gather some feedback

thecoolwinter commented on CodeEdit/Features/QuickOpen/ViewModels/QuickOpenViewModel.swift:
```diff
@@ -47,19 +47,24 @@ final class QuickOpenViewModel: ObservableObject {
                 }
+
+                /// sorts the filtered filePaths with the FuzzySearch
+                let ordertFiles = FuzzySearch.search(query: self.openQuicklyQuery, in: filteredFiles)
```
Spelling error.

thecoolwinter commented on CodeEdit/Features/QuickOpen/ViewModels/FuzzySearch.swift:
```diff
@@ -0,0 +1,207 @@
+}
+
+extension String {
+    /// Adds a new function to the `String` type that searches for all occurrences
```
No need for the "Adds a new" portion of the comment. You can simply describe what the method does.

thecoolwinter reviewed (changes requested):
This is a very in-depth fuzzy search implementation. Well done! I left a few comments and want two more things to be fixed before I can approve this:
- The empty `Icon` file at the root of the project needs to be removed. Looks like it was maybe created by accident.
- It looks like some files are just ignored by the search. The best example I can find is some of the files in the root of the CodeEdit project. I've attached a few screenshots of the search not working for a few files, but working for the README in the same folder.

<img width="1249" alt="Screenshot 2023-05-25 at 3 43 14 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/35942988/59dbc5bd-5bd0-4bde-b432-e22659c0d927">
<img width="1249" alt="Screenshot 2023-05-25 at 3 43 04 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/35942988/b19ee306-dacf-4f0c-bde4-b555cdbabadd">
<img width="1249" alt="Screenshot 2023-05-25 at 3 42 55 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/35942988/d1f7d465-f06b-4198-875b-83b582bddfc6">

activcoding commented:
Thanks for the review. I corrected the spelling error, removed the `Icon` file, removed unnecessary documentation and fixed the bug where the quick search didn't show some files. Everything should work now.

thecoolwinter commented on CodeEdit/Features/QuickOpen/ViewModels/QuickOpenViewModel.swift:
```diff
@@ -42,24 +42,28 @@ final class QuickOpenViewModel: ObservableObject {
                     do {
                         let values = try url.resourceValues(forKeys: [.isRegularFileKey])
-                        return state1 && (values.isRegularFile ?? false)
+                        return (values.isRegularFile ?? false) && !url.absoluteString.contains(".git")
```
Don't exclude the git directory like this. It shouldn't contain any files that might mess up a search and it'll be clear in the search results that they are from the git directory.

austincondiff replied:
At least don't do this manually. We might include a setting at some point that excludes specified things from search results.

activcoding replied:
OK no problem. I just figured the user shouldn't see something like this:
<img width="1119" alt="Screenshot 2023-05-29 at 6 52 05 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/83090745/59f56483-551d-4b58-9962-842ad7e3bf6e">

thecoolwinter commented:
It looks like some files are still being excluded from the search results. For instance I can't find the `World.swift`, `.gitignore`, `.swiftlint.yml` in the CodeEdit folder when using this search.

activcoding commented:
> It looks like some files are still being excluded from the search results. For instance I can't find the `World.swift`, `.gitignore`, `.swiftlint.yml` in the CodeEdit folder when using this search.

When I search all the files show up: 
<img width="1119" alt="Screenshot 2023-05-29 at 6 56 33 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/83090745/f50730e3-e587-4f2c-9221-9250cda47d04">
<img width="1119" alt="Screenshot 2023-05-29 at 7 03 05 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/83090745/a37932ef-6fcf-4c0f-9c86-0d49f203b513">
<img width="1119" alt="Screenshot 2023-05-29 at 6 52 41 PM" src="https://github.com/CodeEditApp/CodeEdit/assets/83090745/0b7ba7c2-d268-4209-893f-8d3aac10b47d">

Wouter01 commented on CodeEdit/Features/QuickOpen/ViewModels/FuzzySearch.swift:
```diff
@@ -0,0 +1,203 @@
+        let sortedResult = filteredResult.sorted { url1, url2 -> Bool in
+            let nameScore1 = score(query: query, url: url1)
+            let nameScore2 = score(query: query, url: url2)
+            if nameScore1 > nameScore2 {
+                return true
+            } else {
+                return false
+            }
```
This can be shorter written as `return nameScore1 > nameScore2`

activcoding replied:
Thanks for the tip 🙏

Wouter01 reviewed (changes requested):
It looks like there's an issue in your code, as it only shows one result, whereas I'd expect it to return 3 results (as seen in the screenshot)

![image](https://github.com/CodeEditApp/CodeEdit/assets/62355975/1db08736-b28b-4fc1-808c-6512e3ba01a2)

I printed out the nameScore in the search function, and it gives the following:
```
8.464783308293057e-59
0.0
0.0
... (others are all zero)
```

Is this really small number normal?

activcoding commented:
Yep, the small numbers are normal. But 0.0 is definitely too small when 3 characters match.
I'll take a look at it tomorrow

activcoding commented:
Hey @Wouter01, 
this morning I tried to reproduce the error but everything works on my machine (see picture). Could you please share the folder, that would help me to understand the error.

<img width="1287" alt="Screenshot 2023-05-30 at 9 54 52 AM" src="https://github.com/CodeEditApp/CodeEdit/assets/83090745/0c55fa81-e3c3-4ea3-8e4e-507f4f073128">

austincondiff commented:
Closing because of #1307.
